### ISSUE TITLE
Known contract violation errors in the GitHub GraphQL API

### ISSUE THREAD

**Issue by wchargin** [sourcecred/sourcecred#996]
This issue documents instances in which the GitHub GraphQL API specifies
that it returns objects of certain types, but in fact returns objects of
the wrong type (and lies about the objects’ actual types).

Depending on what fields exist on the relevant types, queries can also
sometimes crash or return invalid data.

SourceCred developers should know about these because we depend on the
GitHub API. We should try to avoid querying them when possible. This may
not always be possible—we definitely need `Reaction.user`, for instance.
But we can probably avoid requesting `IssueComment.issue` even if it may
be marginally convenient in some case.

Will post as individual comments for direct linking.

Table of contents:

  - [`Reaction.user`]
  - [`GitActor.user`]
  - [`IssueComment.issue`]
  - [`Release.author`]

[`Reaction.user`]: https://github.com/sourcecred/sourcecred/issues/996#issuecomment-435495422
[`GitActor.user`]: https://github.com/sourcecred/sourcecred/issues/996#issuecomment-435495461
[`IssueComment.issue`]: https://github.com/sourcecred/sourcecred/issues/996#issuecomment-435495537
[`Release.author`]: https://github.com/sourcecred/sourcecred/issues/996#issuecomment-456718463

**Comment by wchargin** [comment]
## `Reaction.user`

The [`Reaction` object] has a field `user` of type `User`. In fact,
sometimes the returned objects are of type `Organization`. This can
occur if a user reacts to a post and then converts their account to an
organization: existing reactions are not deleted, and the GraphQL API
returns the wrong result.

Steps to reproduce:

  - Create new accounts **A** and **B**.
  - On account **A**, react to some issue, issue comment, etc.
  - Transform account **A** to an organization owned by **B**.
  - Use the GraphQL API to request the reactions for the post in
    question, and select `nodes { user { __typename id } }`. The result
    will have `__typename: "User"`, but the `id` will correspond to an
    organization ID: re-fetching with `node(id: …) { __typename }` will
    yield `Organization`.

Sample query:

```graphql
{
  reactionNode: node(id: "MDg6UmVhY3Rpb24yMTY3ODkyNQ==") {
    __typename
    ... on Reaction {
      user {
        __typename
        id
        login
      }
    }
  }
  userNode: node(id: "MDEyOk9yZ2FuaXphdGlvbjE3OTUyOTI1") {
    __typename
    id
  }
}
```

```json
{
  "data": {
    "reactionNode": {
      "__typename": "Reaction",
      "user": {
        "__typename": "User",
        "id": "MDEyOk9yZ2FuaXphdGlvbjE3OTUyOTI1",
        "login": "webineh"
      }
    },
    "userNode": {
      "__typename": "Organization",
      "id": "MDEyOk9yZ2FuaXphdGlvbjE3OTUyOTI1"
    }
  }
}
```

**Comment by wchargin** [comment]
## `GitActor.user`

The `GitActor` object has a field `user` of type `User`. In fact,
sometimes the returned user can be a `Bot`.

For example, take the merge commit of this pull request:
<https://github.com/ipfs/js-ipfs-unixfs-engine/pull/122>

Sample query:

```graphql
{
  commitNode: node(id: "MDY6Q29tbWl0NDkwMDgwODk6NjVhNjc1OTJhMWQ3M2VjYjNlN2YzMzRkYmE5NmExMDY0YWQzMjYzMA==") {
    __typename
    ... on Commit {
      author {
        __typename
        user {
          __typename
          id
          login
          url
        }
      }
    }
  }
  authorNode: node(id: "MDM6Qm90MjMwNDAwNzY=") { 
    __typename
    id
    ... on Bot {
      login
      url
    }
  }
}
```

```json
{
  "data": {
    "commitNode": {
      "__typename": "Commit",
      "author": {
        "__typename": "GitActor",
        "user": {
          "__typename": "User",
          "id": "MDM6Qm90MjMwNDAwNzY=",
          "login": "greenkeeper[bot]",
          "url": "https://github.com/greenkeeper%5Bbot%5D"
        }
      }
    },
    "authorNode": {
      "__typename": "Bot",
      "id": "MDM6Qm90MjMwNDAwNzY=",
      "login": "greenkeeper",
      "url": "https://github.com/apps/greenkeeper"
    }
  }
}
```

**Comment by wchargin** [comment]
## `IssueComment.issue`

The `IssueComment` object has a field `issue` of type `Issue!`. But
`IssueComment`s represent comments on both issues and pull requests.
Requesting the `issue` field on an `IssueComment` that belongs to a pull
request yields incorrect results.

Sample query:

```graphql
{
  commentNode: node(id: "MDEyOklzc3VlQ29tbWVudDQzNTEyNjczOQ==") {
    __typename
    ... on IssueComment {
      issue {
        __typename
        id
        url
      }
    }
  }
  pullRequestNode: node(id: "MDExOlB1bGxSZXF1ZXN0MjI3NzAyMDUw") {
    __typename
    id
    ... on PullRequest {
      url
    }
  }
}
```

```json
{
  "data": {
    "commentNode": {
      "__typename": "IssueComment",
      "issue": {
        "__typename": "Issue",
        "id": "MDExOlB1bGxSZXF1ZXN0MjI3NzAyMDUw",
        "url": "https://github.com/sourcecred/sourcecred/pull/975"
      }
    },
    "pullRequestNode": {
      "__typename": "PullRequest",
      "id": "MDExOlB1bGxSZXF1ZXN0MjI3NzAyMDUw",
      "url": "https://github.com/sourcecred/sourcecred/pull/975"
    }
  }
}
```

**Comment by wpank** [comment]
I've done some similar [experimenting here.](https://github.com/casper0001/special-octo-waffle/issues/1)

Take a look if you get the chance. A lot of it is about a user transforming into an organization. 

These are some conclusions I came to when User -> Organization:
- In an issue's comment, the author will become null. There is no way to re-query who made the comment.
- In a Reaction, all the information about a user will remain unchanged. They can be re-queried via a direct node lookup to find their appropriate `__typename`.
- In a Commit within a Pull Request, the author's name field will remain unchanged. The user of the author will change to null. The creator of the commit can be re-queried with the `name` field of the author.

[Github Warns:](https://help.github.com/articles/converting-a-user-into-an-organization/)
```
Warning: Before converting a user into an organization, keep these points in mind:

You will no longer be able to sign into the converted user account.
You will no longer be able to create or modify gists owned by the converted user account.
An organization cannot be converted back to a user.
The SSH keys, OAuth tokens, job profile, and associated user information, will not be transferred to the organization. This is only true for the user account that's being converted, not any of the user account's collaborators.
Any commits made with the converted user account will no longer be linked to that account. The commits themselves will remain intact.
```

They make it sound like everything should be unlinked from a user to the organization it becomes. Basically, the user is "deleted". 

If that's the case, anything with a `null` author or user should be discarded. 

The expected behavior for reactions (like when an account is deleted) should be for the organization's reactions to be deleted as well. 

Since this is not the case with reactions, you can lookup the user with their id, and if their `__typename` is an organization, discard the reaction. 

@wchargin any thoughts on this?

cc: @decentralion

**Comment by wchargin** [comment]
Yep, I did something similar at wchargin/ghost-playground. It sounds
like we have similar naming schemes. :-)

I reported this to GitHub back in November, and their response was
basically, “thanks, good to know; your understanding is correct, strange
things happen sometimes, and we don’t plan to do anything about it”.

> * In a Commit within a Pull Request, the author's name field will
> remain unchanged. The user of the author will change to null. The
> creator of the commit can be re-queried with the `name` field of the
> author.

Can you be a bit more specific here? Do you mean that [`Commit.author`]
is a non-null [`GitActor`] whose `name` is non-null but whose `User` is
null? If so, this is what I’d expect: the `name` and `email` are in the
Git commit itself, and the email no longer resolves to a GitHub user.

[`Commit.author`]: https://developer.github.com/v4/object/commit/#author
[`GitActor`]: https://developer.github.com/v4/object/gitactor/

> If that's the case, anything with a `null` author or user should be
> discarded.

Not sure about this: just because someone deletes their user account
doesn’t mean that all the issues that they’ve created and all the
comments that those issues have received should be deleted from our
records.

> The expected behavior for reactions (like when an account is deleted)
> should be for the organization's reactions to be deleted as well.

Yep, agreed, given that it doesn’t make much sense for a ghost to react
to something.

> Since this is not the case with reactions, you can lookup the user
> with their id, and if their `__typename` is an organization, discard
> the reaction.

Yes, in principle. But where would this logic live, and how would you
formulate this rule in the context of (presumably) a broader mechanism
in Mirror? Please see also the “Alternatives considered” section of #998
for further discussion; this is related to the “Exceptional handling of
exceptional cases” subsection.

**Comment by wpank** [comment]
> Can you be a bit more specific here? Do you mean that Commit.author
is a non-null GitActor whose name is non-null but whose User is
null? If so, this is what I’d expect: the name and email are in the
Git commit itself, and the email no longer resolves to a GitHub user.

Yes, this is correct. The query result for example is: 
<details><summary>Query Result</summary>
<p>

```
{
  "data": {
    "repository": {
      "pullRequests": {
        "nodes": [
          {
            "commits": {
              "nodes": [
                {
                  "commit": {
                    "author": {
                      "name": "casper0004",
                      "__typename": "GitActor",
                      "user": null
                    }
                  }
                }
              ]
            }
          }
        ]
      }
    }
  }
}
```
</p>
</details>

Where the `author`'s `name` remains intact, but the `user` becomes `null`. 

> Not sure about this: just because someone deletes their user account
doesn’t mean that all the issues that they’ve created and all the
comments that those issues have received should be deleted from our
records.

Here I mean that any node with an `author` of `null` should disregarded. The issues they've created and the comments that other people have responded to within those issues should still be kept in our records. 

How exactly are deleted users currently handled?

> But where would this logic live, and how would you
formulate this rule in the context of (presumably) a broader mechanism
in Mirror? Please see also the “Alternatives considered” section of #998
for further discussion; this is related to the “Exceptional handling of
exceptional cases” subsection.

This I'm not exactly sure of.  Intuitively, I can that once a node that is a Reaction is currently being added to the graph, add a subsequent query to github before it gets stored in order to check if it's a valid reaction. But I don't think the code is structured this way. 

I'm not too much of a fan of adding a `Fidelity` field, since I believe this issue is isolated to only reactions, and only when users transform into an organization.

**Comment by wchargin** [comment]
> Yes, this is correct. The query result for example is:

Great; thanks. IMO we should store that response exactly as is: it is a
commit with `author != null`, `author.name = "casper0004"`, and
`author.user = null`. This is indistinguishable from the case where a
repository includes a commit whose email address is not associated with
a GitHub account at all.

> Here I mean that any node with an `author` of `null` should
> disregarded. The issues they've created and the comments that other
> people have responded to within those issues should still be kept in
> our records.

I don’t understand this language: an `Issue` that was authored by a user
who transformed into an organization has an `author` of `null`, but we
both agree that we do not want to disregard the issue node. In any
case, it seems like we agree on the important bits.

> How exactly are deleted users currently handled?

Nothing special happens. In the database, authored entities have `null`
authors. In the RelationalView, the list of authors is empty. Everything
just works.

> I'm not too much of a fan of adding a `Fidelity` field, since I
> believe this issue is isolated to only reactions and only when users
> transform into an organization.

It’s not thus isolated; see [the notes about `GitActor.user`][gau],
which relates to commit authorship (not reactions) and bots (not
organizations, and not necessarily necessarily transformations).

When @decentralion and I discussed this back in November, we were pretty
hesitant to make any assumptions about the limited scope of this
problem. Empirically speaking, we’ve made a tiny number of requests to a
tiny number of repositories, using a small subset of the GraphQL schema
exposed by GitHub. We think that it’s pretty likely that there are other
contract violation errors, and, given GitHub Support’s responses to my
bug reports, they don’t plan to fix them.

Briefly, here is another error case in the GraphQL API that does not fit
into any of the above cases at all: a `PullRequestReviewComment` has a
non-null field `commit` of type `Commit`, but sometimes requesting the
commit for a review comment causes the query to fail and report an
internal error on GitHub’s side. (Specifically, this occurs when the
commit history of a repository has been scrubbed [by contacting GitHub
Support][scrub].

[gau]: https://github.com/sourcecred/sourcecred/issues/996#issuecomment-435495461
[scrub]: https://help.github.com/articles/removing-sensitive-data-from-a-repository/#using-filter-branch

**Comment by wchargin** [comment]
> So, when we get a non-null User object that has an `id`, would it make
> sense to re-query the remote state of the user before saving it to the
> database, using this state as the canonical version?

If I understand what you’re saying: yes, this would make sense.

Note that [we use shallow queries][sq], so when we query for (e.g.) the
author of an issue, we only request that user’s `__typename` and `id`,
and fetch the user’s data on a subsequent query. All that’s stored on
the issue object in our database is the user’s ID (not typename), and
the ID is technically sufficient to fully determine the user when we
want to fetch its content later.

Currently, we _do_ store the typename, so that we don’t have to post
another query just to get the typename before we can fetch the actual
content. But this is just an optimization. The problem is that this
optimization is only sound if GitHub always returns consistent
`__typename`s for a node, which it does not.

So, if I understand correctly, you’re suggesting that we remove this
optimization in certain cases. Don’t store the user’s typename, or
anything else about it other than its ID. Query the remote directly for
its data to ensure that it’s correct.

This is basically what’s proposed in #998.

Does this make sense?

[sq]: https://github.com/sourcecred/sourcecred/issues/622#issuecomment-415166515

> Do you know of any specific examples of this happening that I can play
> around with?

Not on any publicly available repositories, sorry. I can confirm that it
still happens: I have a perfectly valid query that gets the GitHub
GraphQL backend to serve a 500. (Again, I of course recognize that a
query crash is very different than an incorrect response; I present this
only as evidence that it’s not true that GitHub’s API is perfect
everywhere except this one case.)

**Comment by wchargin** [comment]
## `Release.author`

Similar to [`GitActor.user` above][ga-user].

[ga-user]: https://github.com/sourcecred/sourcecred/issues/996#issuecomment-435495461

The `Release` object has a field `author` of type `User`. In fact,
sometimes the returned author can be a `Bot`.

For instance, take this release:
<https://github.com/probot/probot/releases/tag/v7.1.1>

(Thanks to @wpank for the example!)

Sample query:

```graphql
{
  releaseNode: node(id: "MDc6UmVsZWFzZTEyNjAwNjc3") {
    __typename
    ... on Release {
      author {
        __typename
        id
        login
        url
      }
    }
  }
  authorNode: node(id: "MDM6Qm90NDA4MjkwODI=") {
    __typename
    id
    ... on Bot {
      login
      url
    }
  }
}
```

```json
{
  "data": {
    "releaseNode": {
      "__typename": "Release",
      "author": {
        "__typename": "User",
        "id": "MDM6Qm90NDA4MjkwODI=",
        "login": "release-drafter[bot]",
        "url": "https://github.com/release-drafter%5Bbot%5D"
      }
    },
    "authorNode": {
      "__typename": "Bot",
      "id": "MDM6Qm90NDA4MjkwODI=",
      "login": "release-drafter",
      "url": "https://github.com/apps/release-drafter"
    }
  }
}
```

**Comment by wpank** [comment]
@wchargin this may be tangential, but why does the Mirror module exist? Why not use an architecture like [Prisma](https://github.com/prisma/prisma) and [graphql-yoga](https://github.com/prisma/graphql-yoga) to replace the server and ORM layer?

Using those seems much more flexible, maintainable, and modular overall. It also gives us a queryable GraphQL backend at a very low development cost. Plus it would allow for an easier time integrating with [Github Hooks](https://developer.github.com/webhooks/) and [Github Actions](https://developer.github.com/actions/) for a more functional reactive backend. Any thoughts on those?

**Comment by decentralion** [comment]
> @wchargin this may be tangential, but why does the Mirror module exist? Why not use an architecture like [Prisma](https://github.com/prisma/prisma) and [graphql-yoga](https://github.com/prisma/graphql-yoga) or [Hasura](https://hasura.io/) to replace the server and ORM layer?
> 
> Using those seems much more flexible, maintainable, and modular overall. It also gives us a queryable GraphQL backend at a very low development cost. Plus it would allow for an easier time integrating with [Github Hooks](https://developer.github.com/webhooks/) and [Github Actions](https://developer.github.com/actions/) for a more functional reactive backend. Any thoughts on those?

I love the idea of re-using existing code rather than maintaining the mirror, assuming that we can find an existing system that suits our needs. However, it also seems out of scope for this issue. @wpank would you like to open a separate issue for this idea? I'd love to see more details on how we could use an off-the-shelf solution for maintaining and updating a local copy of GitHub's database.

**Comment by wchargin** [comment]
> > > Briefly, here is another error case in the GraphQL API that does not fit
> > > into any of the above cases at all: a PullRequestReviewComment has a
> > > non-null field commit of type Commit, but sometimes requesting the
> > > commit for a review comment causes the query to fail and report an
> > > internal error on GitHub’s side. (Specifically, this occurs when the
> > > commit history of a repository has been scrubbed by contacting GitHub
> > > Support.
> > 
> > Do you know of any specific examples of this happening that I can play
> > around with?
>
> Not on any publicly available repositories, sorry

Given that [the Filecoin repositories are now public][fc], I can provide
more detail. Here’s an example:

```graphql
{
  node(id: "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDIzMDE0Nzk1OQ==") {
    __typename
    ... on PullRequestReviewComment {
      commit {
        id
      }
    }
  }
}
```

```json
{
  "data": {
    "node": null
  },
  "errors": [
    {
      "type": "INTERNAL",
      "message": "Something went wrong while executing your query. Please include [UUID snip] when reporting this issue."
    }
  ]
}
```

The node in question is this pull request review comment:
<https://github.com/filecoin-project/rust-proofs/pull/296#discussion_r230147959>

There’s nothing special about the comment other than that the underlying
commit has been scrubbed.

[fc]: https://filecoin.io/blog/opening-filecoin-project-repos/

**Comment by wchargin** [comment]
The completion of #998 means that we don’t need to keep quite such close
tabs on this anymore. If we find more contract violation errors, we
should still post them to this issue for completeness.
